### PR TITLE
refactor: convert str_byteindex to lua

### PR DESCRIPTION
**Pull request by tris203**
Refactoring to lua of str_byteindex to allow opts to be passed. This will allow `vim.lsp._str_byteindex_enc` to be deprecated.

questions:

where should this function live, is _editor the right place?
where should the opts class be defined

deprectation suggested in #25272

### CODE REVIEW

File: runtime/lua/vim/_editor.lua [hunk 1]
```diff
@@ -627,6 +627,113 @@ function vim.notify(msg, level, opts) -- luacheck: no unused args
   end
 end
 
+-- Helper function to decode the UTF-8 character to a Unicode code point
```
Comment (justinmk) [inline runtime/lua/vim/_editor.lua:630]: never need to mention "helper" in function names. all functions help, else they should not exist.
```suggestion
-- Decodes a UTF-8 character to a Unicode code point.
```

File: runtime/lua/vim/_editor.lua [hunk 2]
```diff
@@ -627,6 +627,113 @@ function vim.notify(msg, level, opts) -- luacheck: no unused args
...
+--- If {opts} is not supplied, it defaults to false (UTF-32). Returns the byte index.
+---
+--- Invalid UTF-8 and NUL is treated like in |vim.str_utfindex()|.
+--- An {index} in the middle of a UTF-16 sequence is rounded upwards to
+--- the end of that sequence.
+---@param s string
+---@param index integer
+---@param opts boolean|nil| { encoding: "utf_8"|"utf-16"|"utf-32" }
+---@return integer
+function vim.str_byteindex(s, index, opts)
```
Comment (justinmk) [inline runtime/lua/vim/_editor.lua:670]: even if the interface continues to live on `vim.x()`, perhaps the implementation could live in `runtime/lua/vim/text.lua` ? (as a private/interval function `text.__foo()`)

File: runtime/lua/vim/_editor.lua [hunk 3]
```diff
@@ -627,6 +627,113 @@ function vim.notify(msg, level, opts) -- luacheck: no unused args
   end
 end
```
Comment (justinmk) [inline runtime/lua/vim/_editor.lua:628]: nit: it's a bit random that the code was added in between vim.notify and vim.notify_once :)

File: runtime/lua/vim/_editor.lua [hunk 4]
```diff
@@ -715,6 +715,124 @@ function vim._on_key(buf, typed_buf)
...
 
+-- Decodes a UTF-8 character to a Unicode code point.
```
Comment (bfredl) [inline runtime/lua/vim/_editor.lua:718]: I don't think we want do duplicate the UTF-8 implementation in lua. A tight, simple "inner loop" like mb_utf_index_to_bytes is the kind of code which makes sense to keep in C.

Doing the argument parsing and validation in lua makes sense, but then it could shell out to a private, smaller, lua wrapper around `mb_utf_index_to_bytes` .

File: runtime/lua/vim/_editor.lua (hunk 2 above)
Comment (tris203) [reply runtime/lua/vim/_editor.lua:763]: I did look at `vim.text` as a home for this. But didn't want to change the namespace to keep this change non-breaking.

Happy to move it with a wrapper if you think that is better (and will be bytecode compiled still)

File: runtime/lua/vim/_editor.lua (hunk 4 above)
Comment (tris203) [reply runtime/lua/vim/_editor.lua:718]: I did it in Lua as thats what I know and kept the whole implementation more tightly coupled.

Happy to do as you suggest and reach to C, but I might need some help with how best to action it

File: runtime/lua/vim/_editor.lua (hunk 2 above)
Comment (justinmk) [reply runtime/lua/vim/_editor.lua:763]: > `_editor.lua` is guaranteed to be "compiled" in the `nvim` binary as lua byte code. I'm not sure if that applies to text.lua .

for this reason it should probably stay where you have it.

File: runtime/lua/vim/_editor.lua [hunk 5]
```diff
@@ -715,6 +715,124 @@ function vim._on_key(buf, typed_buf)
...
+  if index > len then
+    return len
+  end
+
+  if type(opts) == 'boolean' then
+    if opts then
+      opts = { encoding = 'utf-16' }
+    else
+      opts = { encoding = 'utf-32' }
+    end
```
Comment (justinmk) [inline runtime/lua/vim/_editor.lua:778]:
```suggestion
    opts = opts and { encoding = 'utf-16' } or { encoding = 'utf-32' }
```

File: runtime/lua/vim/_editor.lua [hunk 6]
```diff
@@ -715,6 +715,120 @@ function vim._on_key(buf, typed_buf)
...
+---
+--- {opts} can also be supplied as a boolean, true for utf-16 and false for utf-32.
+--- If {opts} is not supplied, it defaults to false (UTF-32). Returns the byte index.
+---
+--- Invalid UTF-8 and NUL is treated like in |vim.str_utfindex()|.
+--- An {index} in the middle of a UTF-16 sequence is rounded upwards to
+--- the end of that sequence.
+---@param s string
+---@param index integer
+---@param opts? boolean|nil| { encoding: "utf_8"|"utf-16"|"utf-32", error?: boolean }
```
Comment (tris203) [inline runtime/lua/vim/_editor.lua:761]:
```suggestion
---@param opts? boolean|nil| { encoding: "utf-8"|"utf-16"|"utf-32", error?: boolean }
```

File: runtime/lua/vim/_editor.lua [hunk 7]
```diff
@@ -715,6 +715,120 @@ function vim._on_key(buf, typed_buf)
...
+---@return integer
+function vim.str_byteindex(s, index, opts)
```
Comment (tris203) [inline runtime/lua/vim/_editor.lua:763]: I should ```vim.validate()``` here

File: runtime/lua/vim/_editor.lua [hunk 8]
```diff
@@ -715,6 +715,120 @@ function vim._on_key(buf, typed_buf)
...
+  end
+
+  if not opts or not opts.error then
+    opts.error = true
+  end
+
+  local valid_encodings = { 'utf-8', 'utf-16', 'utf-32' }
+  if not vim.tbl_contains(valid_encodings, opts.encoding) then
+    error('Invalid encoding: ' .. opts.encoding)
+  end
```
Comment (glepnir) [inline runtime/lua/vim/_editor.lua:788]:
```suggestion
  opts = opts or { encoding = 'utf-32', error = true }
  if type(opts) == 'boolean' then
    opts = opts and { encoding = 'utf-16' } or { encoding = 'utf-32' }
  end

  local encoding = opts.encoding or 'utf-32'
  local valid_encodings = { utf_8 = true, utf_16 = true, utf_32 = true }
  if not valid_encodings[encoding] then
    error('Invalid encoding: ' .. encoding)
  end
```

File: runtime/lua/vim/_editor.lua [hunk 9]
```diff
@@ -715,6 +715,120 @@ function vim._on_key(buf, typed_buf)
   end
 end
 
+-- Decodes a UTF-8 character to a Unicode code point.
+---@private
+---@param s string
+---@return integer
+local function utf_ptr2char(s)
+  local b1 = s:byte(1)
```
Comment (glepnir) [inline runtime/lua/vim/_editor.lua:723]: why not get them in once `local b1, b2, b3, b4 = s:byte(1, 4)`

File: runtime/lua/vim/_editor.lua [hunk 10]
```diff
@@ -715,6 +715,107 @@ function vim._on_key(buf, typed_buf)
...
+    utf16_ptr = utf16_ptr + char_len + 1
+
+    if opts.encoding == 'utf-16' and utf16_char >= index then
+      return utf16_ptr
+    end
+
+    if opts.encoding == 'utf-32' and utf32_char >= index then
+      return utf32_ptr
+    end
+  end
```
Comment (glepnir) [inline runtime/lua/vim/_editor.lua:811]:
```suggestion
  -- Prepare string by removing NUL characters
  local prepared_string = s:gsub('%z', ' ')
  local strlen = vim.fn.strchars(prepared_string)
  local utf16_ptr, utf16_char = 0, 0
  local utf32_ptr, utf32_char = 0, 0
  local j = 0

  -- Traverse the string and calculate pointers for UTF-16 and UTF-32
  while j <= strlen do
    j = j + 1
    local char_len = vim.str_utf_end(s, j)
    local c = utf_ptr2char(s:sub(j, j + char_len))

    if c > 0xFFFF then
      utf16_char = utf16_char + 2
    else
      utf16_char = utf16_char + 1
    end
    utf32_char = utf32_char + 1

    if encoding == 'utf-16' and utf16_char >= index then
      return utf16_ptr + char_len
    elseif encoding == 'utf-32' and utf32_char >= index then
      return utf32_ptr + char_len
    end

    utf32_ptr = utf32_ptr + char_len + 1
    utf16_ptr = utf16_ptr + char_len + 1
  end
```

Comment (tris203) [reply runtime/lua/vim/_editor.lua:811]: I have taken some aspects from this comment to clean the loop up, but I think I need the str_pos() for edge cases around split bytes

Comment (tris203) [reply runtime/lua/vim/_editor.lua]: I have taken a bit of a merged approach, as we did need to track the byte index too for substringing when the byte is split.

I have pushed a separate commit for now, will squash when you have taken a look @glepnir

Comment (glepnir) [reply runtime/lua/vim/_editor.lua]: most LGTM just one thing shouldn't `if s:find('%z') then` do sub ? sub will have memory allocation to generate a new string.

Comment (tris203) [reply runtime/lua/vim/_editor.lua]: i have added a conditional check, so the sub is only done if a NUL is found

Comment (glepnir) [reply runtime/lua/vim/_editor.lua]: LGTM now.

File: runtime/lua/vim/_editor.lua (hunk 4 above)
Comment (tris203) [reply runtime/lua/vim/_editor.lua:718]: As another thought. I could move the string to codepoint lua code to ```vim.text``` and potentially make it public for use it in other places

It is also used in my PR for ```str_utfindex``` #30600

I do think that having the implementation all in pure lua is simpler

File: src/nvim/lua/stdlib.c [hunk 11]
```diff
@@ -696,9 +667,6 @@ void nlua_state_add_stdlib(lua_State *const lstate, bool is_thread)
     // str_utfindex
     lua_pushcfunction(lstate, &nlua_str_utfindex);
     lua_setfield(lstate, -2, "str_utfindex");
-    // str_byteindex
-    lua_pushcfunction(lstate, &nlua_str_byteindex);
-    lua_setfield(lstate, -2, "str_byteindex");
```
Comment (justinmk) [inline src/nvim/lua/stdlib.c:701]: this could be renamed to:

    lua_setfield(lstate, -2, "__str_byteindex");

and then the `vim.str_byteindex` that you defined in `runtime/lua/vim/_editor.lua` can call into this.

Comment (tris203) [reply src/nvim/lua/stdlib.c:701]: have pushed a fixup as a seperate commit, will squash once reviewed

Thanks for the advice/guidance

File: runtime/lua/vim/_editor.lua [hunk 12]
```diff
@@ -715,6 +715,58 @@ function vim._on_key(buf, typed_buf)
...
+
+  opts = opts or { encoding = 'utf-32', error = true }
+  if type(opts) == 'boolean' then
+    opts = opts and { encoding = 'utf-16', error = true } or { encoding = 'utf-32', error = true }
+  end
+  local encoding = opts.encoding or 'utf-32'
+  local valid_encodings = { ['utf-8'] = true, ['utf-16'] = true, ['utf-32'] = true }
+  if not valid_encodings[encoding] then
+    error('Invalid encoding: ' .. encoding)
+  end
```
Comment (lewis6991) [inline runtime/lua/vim/_editor.lua:757]: This function should be fast, however a lot of the code here will add unnecessary overhead for typical call arguments.

Can some of this logic be bypassed if opts isn't provided?

File: runtime/lua/vim/_editor.lua [hunk 13]
```diff
@@ -715,6 +715,58 @@ function vim._on_key(buf, typed_buf)
...
+    end
+    return len
+  end
+
+  opts = opts or { encoding = 'utf-32', error = true }
+  if type(opts) == 'boolean' then
+    opts = opts and { encoding = 'utf-16', error = true } or { encoding = 'utf-32', error = true }
+  end
+  local encoding = opts.encoding or 'utf-32'
+  local valid_encodings = { ['utf-8'] = true, ['utf-16'] = true, ['utf-32'] = true }
```
Comment (lewis6991) [inline runtime/lua/vim/_editor.lua:754]: Pull this constant out as an upvalue.

File: runtime/lua/vim/_editor.lua [hunk 14]
```diff
@@ -68,6 +68,13 @@ vim.log = {
   },
 }
 
+--- @private
+vim.utfs = {
```
Comment (justinmk) [inline runtime/lua/vim/_editor.lua:72]: does this need to be on the public interface?

```suggestion
local utfs = {
```

Comment (tris203) [reply runtime/lua/vim/_editor.lua]: no, have made it local

File: runtime/lua/vim/_editor.lua (hunk 12 above)
Comment (tris203) [reply runtime/lua/vim/_editor.lua:762]: added an early return for a boolean or nil opts

File: runtime/lua/vim/_editor.lua [hunk 15]
```diff
@@ -68,6 +68,12 @@ vim.log = {
   },
 }
 
+local utfs = {
+  ['utf-8'] = 'utf-8',
```
Comment (lewis6991) [inline runtime/lua/vim/_editor.lua:72]: I would make all the values `true` and inline the unnecessary references to `utfs` below.

File: runtime/lua/vim/_editor.lua [hunk 16]
```diff
@@ -715,6 +721,57 @@ function vim._on_key(buf, typed_buf)
...
+--- Invalid UTF-8 and NUL is treated like in |vim.str_utfindex()|.
+--- An {index} in the middle of a UTF-16 sequence is rounded upwards to
+--- the end of that sequence.
+---@param s string
+---@param index integer
+---@param opts? boolean|nil| { encoding: "utf-8"|"utf-16"|"utf-32", error?: boolean }
+---@return integer
+function vim.str_byteindex(s, index, opts)
+  vim.validate({ s = { s, 'string' }, index = { index, 'number' } })
+  local len = #s
```
Comment (lewis6991) [inline runtime/lua/vim/_editor.lua:743]: This line can be moved forward.

File: runtime/lua/vim/_editor.lua [hunk 17]
```diff
@@ -715,6 +721,57 @@ function vim._on_key(buf, typed_buf)
...
+  if index == 0 then
+    return 0
+  end
+
+  if type(opts) == 'boolean' or not opts then
+    return vim.__str_byteindex(s, index, opts)
+  end
+
```
Comment (lewis6991) [inline runtime/lua/vim/_editor.lua:751]: If you make the above change then you can move vim.validate here. There is also an alternate version of vim.validatenwhich is faster as it uses less tables.

File: test/functional/lua/vim_spec.lua [hunk 18]
```diff
@@ -319,20 +319,99 @@ describe('lua stdlib', function()
...
+        'return vim.str_byteindex(_G.test_text, 99999, { encoding = "utf-16", error = false})'
+      )
+    )
+    eq(
+      indices8[#indices8],
+      exec_lua(
+        'return vim.str_byteindex(_G.test_text, 99999, { encoding = "utf-8", error = false})'
+      )
+    )
+    eq(2, exec_lua('return vim.str_byteindex("é", 2, { encoding = "utf-16", error = false})'))
```
Comment (tris203) [inline test/functional/lua/vim_spec.lua:414]: With the C implementation, this will still error (even though error is false) as ```#str = 2``` which means that index goes out of range and throws from the C side

File: runtime/lua/vim/_editor.lua [hunk 19]
```diff
@@ -715,6 +721,57 @@ function vim._on_key(buf, typed_buf)
...
+  if index > len then
+    if opts.error then
+      error('index out of range')
+    end
+    return len
+  end
+
+  opts = opts or { encoding = 'utf-32', error = true }
```
Comment (lewis6991) [inline runtime/lua/vim/_editor.lua:759]: I can't see where `opts.error` is used.

**Comment by tris203** [comment]
Okay, following my comment here about edge cases, and adding a failing test.
I reverted back to the lua implementation for testing, 

I have taken all of the comments above onboard, and have implemented several performance improvements whilst still handling the edge cases around indexes and byte length and bfredls comments from Matrix around UTF-8 indices.

1. Use faster vim validate
2. Drop all string manipulations and use byte indicies
3. Drop the vim script call out for charlen and track the loop via bytes

Using the below test code to run the loop 10 million times looking for the last character (worst case)
```lua
local str = "jjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjj"
local x = 10000000
local total = 0
while x > 0 do
  x = x - 1
local start = vim.uv.hrtime()
vim.str_byteindex(str, 120, { encoding = "utf-16" })
local stop = vim.uv.hrtime()
local elapsed = stop - start
  total = total + elapsed
end
local avg = total / 10000000
print(avg / 1000000)
```

current master:
0.0004058819099ms

lua implementation:
0.0003157661928

so it is actually faster, and handles more edge cases and in my arguably easier to read/understand

**Comment by tris203** [comment]
Is there any feedback on the latest findings. Apologies if I haven't made it clear.

However, I think the lua "wrapper" around the existing C function can't deal with the edge cases sufficiently without pcalls, which isn't ideal.

My findings above seem to indicate that the performance in Lua is about the same/if not faster for most cases

File: runtime/doc/lua.txt [hunk 20]
```diff
@@ -1772,6 +1756,29 @@ vim.schedule_wrap({fn})                                  *vim.schedule_wrap()*
...
+    the byte index.
+
+    Invalid UTF-8 and NUL is treated like in |vim.str_utfindex()|. An {index}
+    in the middle of a UTF-16 sequence is rounded upwards to the end of that
+    sequence.
+
+    Parameters: ~
+      • {s}      (`string`)
+      • {index}  (`integer`)
+      • {opts}   (`boolean?|{ encoding: "utf-8"|"utf-16"|"utf-32", error?: boolean }?`)
```
Comment (lewis6991) [inline runtime/doc/lua.txt:1777]: Is there a reason of having a specific `error` opt instead of using `pcall`?

Comment (tris203) [reply runtime/doc/lua.txt:1777]: the idea was to avoid pcalls and handle the errors gracefully to replace ```vim.lsp.util._str_byteindex_enc()```

Comment (lewis6991) [reply runtime/doc/lua.txt:1777]: To me handling the error gracefully means detecting what the error is and acting appropriately. From what I can tell this change just ignores the error completely. I'm not sure how that is more graceful than using `pcall`?

Comment (lewis6991) [reply runtime/doc/lua.txt:1777]: @clason wdyt?

Comment (justinmk) [reply runtime/doc/lua.txt:1777]: > The inspiration for the interface was from the recent `get_parser()` changes

The `error` flag added to get_parser is a temporary workaround. See `:help dev-error-patterns` for recommended ways to pass errors/failure.

Comment (tris203) [reply runtime/doc/lua.txt:1777]: > > The inspiration for the interface was from the recent `get_parser()` changes
> 
> The `error` flag added to `get_parser` is a temporary workaround, probably not relevant here. Instead see `:help dev-error-patterns` for recommended ways to pass errors/failure.

I have read that documentation. The point of the flag is essentially choosing between behaviour 3 or a fallback value.

I could go with option 1, and provide
```i, nil``` in a success case and ```#s, "index out of range"``` in a failure case. But given that the function throws at the moment, there may be other callers depending on the throw behaviour

Comment (justinmk) [reply runtime/doc/lua.txt:1777]: > The point of the flag is essentially choosing between behaviour 3

The flag isn't a pattern that we want to continue. So unless it's a temporary bridge that we can't avoid, I'd suggest just throwing an error if that's how it was already done and we think changing that would break consumers.

Comment (tris203) [reply runtime/doc/lua.txt:1777]: > > The point of the flag is essentially choosing between behaviour 3
> 
> The flag isn't a pattern that we want to continue. So unless it's a temporary bridge that we can't avoid, I'd suggest just throwing an error if that's how it was already done and we think changing that would break consumers.

we need to be able to suppress the throw and provide a fallback to be able to replace the ```vim.lsp.util._str_byteindex_enc()```

What do you think about a fallback flag? I am trying to think of where we have a similar pattern for precedence

Comment (tris203) [reply runtime/doc/lua.txt:1777]: The closest I can think of is an ```on_error``` callback function that would be similar to ```on_init``` or ```on_attach```

Comment (lewis6991) [reply runtime/doc/lua.txt:1777]: What is the problem with using `pcall`?

Comment (tris203) [reply runtime/doc/lua.txt:1777]: > What is the problem with using `pcall`?

The motivation for rewriting this was part of #25272 to ultimately be able to retire ```_str_byteindex_enc```
We don't want to have to call every LSP call to ```str_byteindex``` as that leads to overhead in callers code and  can mask errors.

``'
                                                          *lua-error-handling*
Lua functions may throw |lua-errors| for exceptional (unexpected) failures,
which you can handle with |pcall()|.
```

I would argue that an index being out of range isn't exceptional enough to throw

Comment (lewis6991) [reply runtime/doc/lua.txt:1777]: > and can mask errors.

What errors? The only error this can really throw is argument type or out of index.

Out of index seems exceptional to me. Using pcall seems like the most ideomatic route. This function just isn't complicated enough to justify doing anything else.

Comment (tris203) [reply runtime/doc/lua.txt:1777]: > > The motivation for rewriting this was part of #25272
> 
> I can't find anything in that issue that specifically mentions anything about avoiding pcall. All it says is "Consider updating those functions to accept an encoding instead."
> 

But there is a bounds check/fallback in the existing ```_str_byteindex_enc``` check here
https://github.com/neovim/neovim/blob/8ef3dd3afa20024f439d8b7a69f316069c1909d3/runtime/lua/vim/lsp/util.lua#L174-L178

Additionally, the current bounds check isn't sufficient if you take the below code
```lua
str = '世e'

vim.print(vim.lsp.util._str_byteindex_enc(str, 1, 'utf-16'))
vim.print(vim.lsp.util._str_byteindex_enc(str, 2, 'utf-16'))
-- vim.print(vim.lsp.util._str_byteindex_enc(str, 3, 'utf-16'))
-- vim.print(vim.lsp.util._str_byteindex_enc(str, 4, 'utf-16'))
vim.print(vim.lsp.util._str_byteindex_enc(str, 5, 'utf-16'))
```

The commented lines will currently throw, the others will return 2, 4, and 4 (and any index above 4 would return 4 too)

so we would need to be able to use that bounds check in a higher level implementation too or we would just need another wrapper function

Comment (lewis6991) [reply runtime/doc/lua.txt:1777]: The functions is called like 4 times. Why not just pull the bounds checking outside? Is it even completely necessary in each cace?

Comment (tris203) [reply runtime/doc/lua.txt:1777]: I like the idea of stopping it throwing and moving it to a ```retval, errmsg``` return

Then the lsp code can just use 
```lua
vim.str_byteindex(string, 10, { encoding = 'utf-16' }) or #string
```

But is that considered breaking for a function that used to throw previously

Comment (lewis6991) [reply runtime/doc/lua.txt:1777]: Yes. Please just use `pcall`, or just throw the error i the LSP code.

**Comment by justinmk** [comment]
Note that https://github.com/neovim/neovim/pull/30735 is a more incremental approach to this, which should probably be merged first, then this PR can be rebased on it, if needed.

**Comment by justinmk** [comment]
Do we still need this?

**Comment by tris203** [comment]
> Do we still need this?

It is faster than the current implementation.
I think Lewis found it was 15% faster, but then there is some discussion as to where we want the code to be

But the flexibility is already provided

**Comment by tris203** [comment]
superseded by #30600
